Cache is_projective_planar() method for graphs

orlitzky opened the pull request:
The main example of this (the Petersen graph) is taking over six minutes on the CI, and even longer on my machine. We already know the answer when the graph is constructed, so by making it a `@cached_method`, the method becomes effectively instantaneous.

We still want to test the computation, but that can be done via pytest (and now is) rather than in a user-facing example.

orlitzky committed:
sage.graphs.graph: make is_projective_planar() a @cached_method

The is_projective_planar() function can be very slow. The first
example in the documentation is the Petersen graph, on which the
method is taking over six minutes to arrive at the answer (True).
We mark it as a @cached_method, since that will allow us to pre-
fill the cache for graphs that are known to be projective planar,
saving the user a lot of time.

orlitzky committed:
sage.graphs.generators.smallgraphs: start caching is_projective_planar()

The Petersen graph is known to be projective planar, so we can
pre-fill the cache with "True" when it is constructed.

orlitzky committed:
sage.graphs.is_projective_planar_test: new is_projective_planar() pytests

To ensure that is_projective_planar() still works in the absense of
the cache (now that it is a @cached_method), we add a new pytest file.
For the moment it only tests the Petersen graph, since that is the
only graph with the method initially cached.

orlitzky committed:
src/sage/graphs/meson.build: add new file is_projective_planar_test.py

Ultimately we may not want to install these test files, but for now
the CI is complaining about it. We can decide later on.

dcoudert commented:
A simpler solution is to use a different graph, for instance:
```sage
sage: G = graphs.CompleteGraph(5)
sage: %time G.is_projective_planar()
CPU times: user 1.16 ms, sys: 16 µs, total: 1.18 ms
Wall time: 1.18 ms
True
sage: G = graphs.CompleteGraph(6)
sage: %time G.is_projective_planar()
CPU times: user 1.15 ms, sys: 51 µs, total: 1.2 ms
Wall time: 1.21 ms
True
sage: G = graphs.CompleteGraph(7)
sage: %time G.is_projective_planar()
CPU times: user 124 ms, sys: 2.01 ms, total: 126 ms
Wall time: 125 ms
False
```
and to mark the test for the Petersen Graph as `# not tested (6 min)`

There is a typo `Peterson` $\to$ `Petersen` in `graph.py`

orlitzky committed:
src/sage/graphs: move is_projective_planar() tests to a subdirectory

A simple `git mv` with an update of the associated meson.build files.

orlitzky committed:
meson.build: create __init__.py for sage.graphs.tests, too

The "tests" subdirectory is installed, so make it a package for
consistency. Add the generated __init__.py to .gitignore while we're
at it.

orlitzky commented:
Done, just need to wait for the CI to confirm that it still gets run from the subdirectory.

dcoudert commented:
This seems ok. I tested it on my laptop.
I'm not sure the failing doctests are related, but please check.

dcoudert commented:
The failing CIs seems related to #41000.

dcoudert reviewed (approved):
LGTM.

orlitzky commented:
Ok, thank you!

tornaria commented:
This adds a test that takes 5+ min and it seems to be now (10.9) always run after doctests (I'm just running `python -m sage.doctest ...`). I believe 5m is not even acceptable for `--long`, but this pytest stuff is (a) always run regardless of `--long` or not and (b) is not even run in parallel.

@orlitzky @tobiasdiez is there something I am missing?

orlitzky commented:
Sorry, this is a case where two independently beneficial changes produced a bad result.

In this PR, the method *itself* used to take 5+ minutes, so the call to `is_projective_planar()` in the test took 5+ minutes as well. I made the method fast by caching it, and migrated the original test (for the correctness of the cached value) to pytest because, yes, obviously 5+ minutes is too long for a doctest. There were constant warnings about it. So far so good; the method got faster, and we killed a slow doctest.

But then IIRC I also reviewed the PR to run pytest by default having completely forgotten about this, creating the current situation where extremely long-running tests can get run without --long. Pytest supports arbitrary "markers" for tests that we can use to mark tests as "long", "super-long", etc. so there is a good, long-term, principled solution to this problem. But maybe in the meantime we should run pytest only with --long? (Tobias?)

dcoudert commented:
An alternative is to mark this method as "not tested, too long (6 min)" ?

orlitzky commented:
> An alternative is to mark this method as "not tested, too long (6 min)" ?

It's nice to have the algorithm tested, the problem in this case is that I moved the test out of the documentation and into a separate pytest file... and then later `sage -t` was made to run pytest on those files by default.

We could comment the pytest file to achieve the same thing, but if it's not an emergency (I've got three or four other PRs that I've promised to review), I think marking this test "extra-long" and then changing `sage -t` to ignore those is a more consistent approach. Everyone is already familiar with the "long time" tags, and if "extra-long" tests are only run when pytest is invoked directly, that lets us test them on the CI but not when users innocently run `sage -t`.

dcoudert commented:
Indeed an "extra-long" tag would be nice.

tobiasdiez commented:
How about adding a `long` marker for pytest, marking this test with it, and then passing the `--long` flag to pytest from `sage -t`?

tornaria commented:
> How about adding a `long` marker for pytest, marking this test with it, and then passing the `--long` flag to pytest from `sage -t`?

Sure, but please also add an "extra-long" one that is not run with `--long`. There are important tests behind `--long` but a 5 minute test doesn't belong there.

Incidentally, the Petersen graph has 10 vertices and 15 edges, right? Why does it take 5 minutes to check anything with that graph? Even the exponential 2^15 is not that large.

tornaria commented:
Also btw, even if I install pytest-xdist, the pytest is run in a single thread.

It is a bit sad that testing 3854 files using 8 threads (using a normal desktop cpu, i7-9700 so not even new) takes:
```
Total time for all tests: 1022.8 seconds
    cpu time: 5825.2 seconds
    cumulative wall time: 6192.6 seconds
```
and then the pytest step takes:
```
================== 94 passed, 2 skipped in 545.61s (0:09:05) ===================
```

Doctesting the same 3854 files using `--long` takes about twice (2057.4 seconds), so adding 545 seconds is a big overhead even for long testing.

EDIT: ...and more than half the 545 seconds is testing planarity of a 10 vertex graph...

dcoudert commented:
> Incidentally, the Petersen graph has 10 vertices and 15 edges, right? Why does it take 5 minutes to check anything with that graph? Even the exponential 2^15 is not that large.

Method `is_projective_planar` checks if the graph contains any of the 35 forbidden minors. Searching for a minor is done with linear programming and it is slow (in particular with glpk). 
There should be a faster method for checking if a graph is projective planar. I don't know the results on this topic.
